### DISCUSSION TITLE
Wireguard integration

### DISCUSSION THREAD

**Discussion by emanuele-f** [emanuele-f/PCAPdroid#397] [Ideas]
PCAPdroid should give users the ability to use a Wireguard tunnel while PCAPdroid is running, with the ability to start/stop the tunnel at runtime.

Here are possible ways to make this possible:

### EveryProxy + Wireguard app + work profile

[By installing EveryProxy and Wireguard in a work profile](https://github.com/emanuele-f/PCAPdroid/issues/260#issuecomment-1275238169), and setting up a SOCKS5 connection to EveryProxy in PCAPdroid, it's currently possible to use PCAPdroid and Wireguard at the same time. However, this not optimal for the following reasons:

- it requires the use of work profiles, which are not practical to configure and use
- in this mode, Wireguard will create a new VPNService, with additional overhead and battery consumption
- manual steps are required to control this setup, e.g. to stop/start the tunnel in the work profile

### Embed Wireguard

As explained in the [embedding guide](https://www.wireguard.com/embedding), it's possible to embed Wireguard in PCAPdroid, the approach used by Rethink DNS. The pros are that we can have the best performance and control. The cons are that:

- PCAPdroid needs to reimplement the Wireguard app UI, allowing to import and create configurations. However, on an initial implementation, this could be limited to import-only
- The user needs to maintain two different configurations, one in PCAPdroid, the other in Wireguard, and control the two apps differently

### Implement a SOCKS5 server interface in the official Wireguard app

If Wireguard provided a SOCKS5 server interface, PCAPdroid or any other apps could easily integrate with it, and let users use the offical app for any kind of configuration. This is the most flexible approach. Note: there is an open feature request for this in wgtunnel, https://github.com/zaneschepke/wgtunnel/issues/87

Moreover in the future, if mitmproxy implements the ability to connect to an upstream SOCSK5 server (https://github.com/mitmproxy/mitmproxy/issues/211), this would allow users to use the tunnel for decrypted connections as well.

### Implement a SOCKS5 server interface in Rethink DNS

If Rethink DNS provided a SOCKS5 server interface, it would be possible for PCAPdroid to use it as a Wireguard and DoT provider. This only works if Rethink DNS is the final app of the chain.

**Comment by anpic** [comment]
> If Rethink DNS provided a SOCKS5 server interface, it would be possible for PCAPdroid to use it as a Wireguard and DoT provider.

I'm sure they themselves have a very vague understanding of how their system works. There are other doubts. I wouldn't rely on them. Although perhaps they already have almost everything for this.
But it's unclear how Wireguard has earned such a favorable attitude compared to OpenVPN

**Comment by ghost** [comment]
Great idea if can be successfully executed. It would also be fundamental for people to use the app as a firewall and don't want to switch off the firewall just to use a VPN.

**Comment by kevin0t** [comment]
An opensource replacement for everyproxy is [singbox](https://github.com/SagerNet/sing-box) which i found more reliable. Also has many other features which i am yet to explore fully 
For an updated version of the setup guide for vpn+ firewall  refer [link](https://itsignacioportal.github.io/firewall-and-vpn-without-root-on-android/) . It uses rethink dns + singbox. But we can idealy use pcapdroid in place of rethink but only catch is regarding dns queries and udp requests.
